core/src/SYCL/Kokkos_SYCL_Parallel_Reduce.hpp

```diff
@@ -575,65 +575,40 @@ class ParallelReduce<CombinedFunctorReducerType,
             m_space.impl_internal_space_instance()->m_mutexScratchSpace) {}
 
  private:
-  template <typename PolicyType, typename CombinedFunctorReducerWrapper>
+  template <typename CombinedFunctorReducerWrapper>
   sycl::event sycl_direct_launch(
-      const PolicyType& policy,
       const CombinedFunctorReducerWrapper& functor_reducer_wrapper,
       const sycl::event& memcpy_event) const {
     // Convenience references
     Kokkos::Experimental::Impl::SYCLInternal& instance =
         *m_space.impl_internal_space_instance();
     sycl::queue& q = m_space.sycl_queue();
 
-    const typename Policy::index_type nwork = m_policy.m_num_tiles;
-    const typename Policy::index_type block_size =
-        std::pow(2, std::ceil(std::log2(m_policy.m_prod_tile_dims)));
-
-    const sycl::range<1> local_range(block_size);
-    // REMEMBER swap local x<->y to be conforming with Cuda/HIP implementation
-    const sycl::range<1> global_range(nwork * block_size);
-    const sycl::nd_range<1> range{global_range, local_range};
-
-    const size_t wgroup_size = range.get_local_range().size();
-    size_t size              = range.get_global_range().size();
-    const auto init_size =
-        std::max<std::size_t>((size + wgroup_size - 1) / wgroup_size, 1);
+    const typename Policy::index_type n_tiles = m_policy.m_num_tiles;
     const unsigned int value_count =
-        m_functor_reducer.get_reducer().value_count();
-    const auto results_ptr =
-        static_cast<sycl::device_ptr<value_type>>(instance.scratch_space(
-            sizeof(value_type) * std::max(value_count, 1u) * init_size));
-    sycl::global_ptr<value_type> device_accessible_result_ptr =
-        m_result_ptr_device_accessible ? m_result_ptr : nullptr;
-    auto scratch_flags = static_cast<sycl::device_ptr<unsigned int>>(
-        instance.scratch_flags(sizeof(unsigned int)));
+        std::max(m_functor_reducer.get_reducer().value_count(), 1u);
+    sycl::device_ptr<value_type> results_ptr;
 
     sycl::event last_reduction_event;
 
-    // If size<=1 we only call init(), the functor and possibly final once
-    // working with the global scratch memory but don't copy back to
-    // m_result_ptr yet.
-    if (size <= 1) {
+    // If n_tiles==0 we only call init() and final() working with the global
+    // scratch memory but don't copy back to m_result_ptr yet.
+    if (n_tiles == 0) {
       auto parallel_reduce_event = q.submit([&](sycl::handler& cgh) {
 #ifndef KOKKOS_IMPL_SYCL_USE_IN_ORDER_QUEUES
         cgh.depends_on(memcpy_event);
 #else
         (void)memcpy_event;
 #endif
+        results_ptr = static_cast<sycl::device_ptr<value_type>>(
```

Comment:
Do you actually need the cast?  Isn't the converting constructor implicit?

Reply:
Yes, I need it.
```
/opt/intel/oneapi/compiler/2023.0.0/linux/bin-llvm/../include/sycl/multi_ptr.hpp:78:14: note: candidate function not viable: no known conversion from 'multi_ptr<void, [...]>' to 'const multi_ptr<int, [...]>' for 1st argument
  multi_ptr &operator=(const multi_ptr &) = default;
             ^
/opt/intel/oneapi/compiler/2023.0.0/linux/bin-llvm/../include/sycl/multi_ptr.hpp:79:14: note: candidate function not viable: no known conversion from 'multi_ptr<void, [...]>' to 'multi_ptr<int, [...]>' for 1st argument
  multi_ptr &operator=(multi_ptr &&) = default;
             ^
/opt/intel/oneapi/compiler/2023.0.0/linux/bin-llvm/../include/sycl/multi_ptr.hpp:89:14: note: candidate function template not viable: no known conversion from 'sycl::device_ptr<void>' (aka 'multi_ptr<void, access::address_space::ext_intel_global_device_space>') to 'pointer_t' (aka '__global int *') for 1st argument
  multi_ptr &operator=(pointer_t pointer) {
             ^
/opt/intel/oneapi/compiler/2023.0.0/linux/bin-llvm/../include/sycl/multi_ptr.hpp:95:14: note: candidate function not viable: no known conversion from 'sycl::device_ptr<void>' (aka 'multi_ptr<void, access::address_space::ext_intel_global_device_space>') to 'int *' for 1st argument
  multi_ptr &operator=(ElementType *pointer) {
             ^
/opt/intel/oneapi/compiler/2023.0.0/linux/bin-llvm/../include/sycl/multi_ptr.hpp:102:14: note: candidate function not viable: no known conversion from 'sycl::device_ptr<void>' (aka 'multi_ptr<void, access::address_space::ext_intel_global_device_space>') to 'std::nullptr_t' for 1st argument
  multi_ptr &operator=(std::nullptr_t) {
             ^
```

Reply:
Ah it is an int* instead of void* so I suppose you need a cast anyway